### ISSUE TITLE
major refactors 

### ISSUE THREAD

**Issue by cryptix** [cryptoscope/ssb#48]
## turn luigi into a pattern

Wanted to document this for a while. This is one of the core reason why go-ssb is so wasteful with RAM.

We use the [luigi](https://github.com/cryptoscope/luigi) package for a lot data processing between the offset-database (margaret) and our muxrpc implementation.

It is _functional_ but wastes a lot of memory since every single message get sent through a channel inside `interface{}`. This means that the messages all end up on the heap as single-use things and then need to be cleaned up again.

One further complication is that we use luigi streams as data transformations (i.e. bytes from the connection > muxrpc frame > validation process), these also involve (un)packing empty interfaces (and escaping to the heap) even though most of these processes could be tight loops and re-using buffers

**Comment by cryptix** [comment]
# margaret does too much

Another problem is that margaret (our flume-like thingy) also does object decoding and exposes it through empty interfaces, which makes it hard to re-use buffers from previous reads, if all you want is pipe out a bunch of messages, like for replication.

I'm currently dabbling around a version of margaret that just exposes `[]byte` or `io.Reader`s, leaving the marshaling up to the API consumer.

This together with a muxrpc version that gives the user access to reading and writing muxrpc frames without the luigi piping issues outlined above, should result in a pub code that's producing much less overhead.

Currently it's easily a couple hundred megabytes per connection.

**Comment by freight-stack** [comment]
How about merging the Luigi and Margaret packages into this project?  It is easier then to iterate and refactor the code base.  We can pull them into single packages again when we are happy with their API.

**Comment by cryptix** [comment]
Yes! Should probably `internal`ize all these  things to make this clearer.

**Comment by cryptix** [comment]
I'm experimenting with making copies of `luigi.Broadcast` with concrete types on [the new rooms implementation](https://github.com/ssb-ngi-pointer/gossb-rooms). Much more readable code..!
